File: desktop/tests/Qlipq.Ffmpeg.Tests/ArgsTests.cs
using Qlipq.Core;
using Qlipq.Ffmpeg;
using Xunit;

namespace Qlipq.Ffmpeg.Tests;

public class ArgsTests
{
    private static List<string> Args(EditSpec spec, Action<BuildExportOptions>? extra = null)
    {
        var opts = new BuildExportOptions { InputPath = "in.mkv", OutputPath = "out.mp4", Spec = spec };
        extra?.Invoke(opts);
        return ExportArgs.BuildExportArgs(opts);
    }

    private static AudioTrackSpec Track(int index, bool enabled, double volume) =>
        new() { Index = index, Enabled = enabled, Volume = volume };

    [Fact]
    public void TrimOnlyDefaultsToFastStreamCopy()
    {
        var outArgs = Args(new EditSpec
        {
            Trim = new TrimSpec { StartSec = 5, EndSec = 12.5 },
            AudioTracks = [Track(0, true, 1)],
        });
        Assert.Equal(
            ["-y", "-ss", "5.000", "-i", "in.mkv", "-t", "7.500", "-map", "0:v:0", "-map", "0:a:0",
                "-c:v", "copy", "-c:a", "copy", "out.mp4"],
            outArgs);
    }

    [Fact]
    public void ForcedReencodeOnTrimOnlyReencodesVideoCopiesAudio()
    {
        var outArgs = Args(
            new EditSpec { Trim = new TrimSpec { StartSec = 0, EndSec = 10 }, AudioTracks = [Track(0, true, 1)] },
            o => o.Reencode = true);
        Assert.Contains("-c:v", outArgs);
        Assert.Contains("libx264", outArgs);
        Assert.Contains("-c:a copy", string.Join(" ", outArgs));
        Assert.DoesNotContain("-filter_complex", outArgs);
    }

    [Fact]
    public void CropBuildsFilterGraphAndReencodesVideo()
    {
        var outArgs = Args(new EditSpec
        {
            Crop = new CropSpec { X = 100, Y = 50, Width = 1280, Height = 720 },
            AudioTracks = [Track(0, true, 1)],
        });
        var joined = string.Join(" ", outArgs);
        Assert.Contains("-filter_complex [0:v:0]crop=1280:720:100:50[vout]", joined);
        Assert.Contains("-map [vout]", joined);
        Assert.Contains("-map 0:a:0", joined);
        Assert.Contains("-c:v libx264", joined);
        Assert.Contains("-c:a copy", joined);
    }

    [Fact]
    public void AudioVolumeChangeReencodesAudioViaFilterCopiesVideo()
    {
        var outArgs = Args(new EditSpec
        {
            AudioTracks = [Track(0, true, 0.5), Track(1, true, 1)],
        });
        var joined = string.Join(" ", outArgs);
        Assert.Contains("-filter_complex [0:a:0]volume=0.5[aout0]", joined);
        Assert.Contains("-map 0:v:0", joined);
        Assert.Contains("-map [aout0]", joined);
        Assert.Contains("-map 0:a:1", joined);
        Assert.Contains("-c:v copy", joined);
        Assert.Contains("-c:a aac -b:a 192k", joined);
    }

    [Fact]
    public void CropPlusVolumeCombinesVideoAndAudioFilters()
    {
        var outArgs = Args(new EditSpec
        {
            Crop = new CropSpec { X = 0, Y = 0, Width = 640, Height = 480 },
            AudioTracks = [Track(0, true, 2)],
        });
        var joined = string.Join(" ", outArgs);
        Assert.Contains("[0:v:0]crop=640:480:0:0[vout];[0:a:0]volume=2[aout0]", joined);
        Assert.Contains("-c:v libx264", joined);
        Assert.Contains("-c:a aac", joined);
    }

    [Fact]
    public void DisablingAllAudioYieldsAnAndNoAudioCodec()
    {
        var outArgs = Args(new EditSpec
        {
            AudioTracks = [Track(0, false, 1), Track(1, false, 1)],
        });
        Assert.Contains("-an", outArgs);
        Assert.DoesNotContain("-c:a", outArgs);
    }

    [Fact]
    public void ProgressFlagAppendsMachineReadableProgress()
    {
        var outArgs = Args(new EditSpec { AudioTracks = [Track(0, true, 1)] }, o => o.Progress = true);
        Assert.Contains("-progress pipe:1 -nostats", string.Join(" ", outArgs));
    }

    [Fact]
    public void CustomEncoderOptionsAreHonoured()
    {
        var outArgs = Args(
            new EditSpec
            {
                Crop = new CropSpec { X = 0, Y = 0, Width = 100, Height = 100 },
                AudioTracks = [Track(0, true, 0)],
            },
            o =>
            {
                o.Video = new VideoEncodeOptions { Codec = "libx265", Crf = 28, Preset = "fast" };
                o.Audio = new AudioEncodeOptions { Codec = "libopus", Bitrate = "96k" };
            });
        var joined = string.Join(" ", outArgs);
        Assert.Contains("-c:v libx265 -preset fast -crf 28", joined);
        Assert.Contains("-c:a libopus -b:a 96k", joined);
    }

    [Fact]
    public void MetadataStampsEntriesBeforeOutput()
    {
        var outArgs = Args(
            new EditSpec { AudioTracks = [Track(0, true, 1)] },
            o => o.Metadata = [new("game", "Deadlock")]);
        var joined = string.Join(" ", outArgs);
        Assert.Contains("-metadata game=Deadlock", joined);
        Assert.True(outArgs.IndexOf("-metadata") < outArgs.IndexOf("out.mp4"));
    }

    [Fact]
    public void FrameRateChangeReencodesAndEmitsRWithoutFilter()
    {
        var outArgs = Args(
            new EditSpec { AudioTracks = [Track(0, true, 1)] },
            o => o.Video = new VideoEncodeOptions { Fps = 30 });
        var joined = string.Join(" ", outArgs);
        Assert.Contains("-c:v libx264", joined);
        Assert.Contains("-r 30", joined);
        Assert.DoesNotContain("-filter_complex", outArgs);
        Assert.Contains("-c:a copy", joined);
    }

    [Fact]
    public void DownscaleBuildsScaleFilterAndReencodes()
    {
        var outArgs = Args(
            new EditSpec { AudioTracks = [Track(0, true, 1)] },
            o => o.Video = new VideoEncodeOptions { ScaleHeight = 720 });
        var joined = string.Join(" ", outArgs);
        Assert.Contains("-filter_complex [0:v:0]scale=-2:720[vout]", joined);
        Assert.Contains("-map [vout]", joined);
        Assert.Contains("-c:v libx264", joined);
    }

    [Fact]
    public void CropAndDownscaleComposeIntoOneFilterChain()
    {
        var outArgs = Args(
            new EditSpec { Crop = new CropSpec { X = 0, Y = 0, Width = 1920, Height = 1080 }, AudioTracks = [] },
            o => o.Video = new VideoEncodeOptions { ScaleHeight = 720 });
        Assert.Contains("[0:v:0]crop=1920:1080:0:0,scale=-2:720[vout]", string.Join(" ", outArgs));
    }

    [Fact]
    public void BitrateRateControlUsesBvInsteadOfCrf()
    {
        var outArgs = Args(
            new EditSpec { AudioTracks = [Track(0, true, 1)] },
            o =>
            {
                o.Reencode = true;
                o.Video = new VideoEncodeOptions { BitrateKbps = 6000 };
            });
        var joined = string.Join(" ", outArgs);
        Assert.Contains("-b:v 6000k", joined);
        Assert.DoesNotContain("-crf", joined);
    }

    private static readonly MediaInfo Media = new()
    {
        DurationSec = 60,
        Width = 2560,
        Height = 1440,
        VideoCodec = "h264",
        Fps = 60,
        AudioStreams = [],
    };

    private static OutputSettings Settings(Action<OutputSettings> over)
    {
        var s = new OutputSettings();
        over(s);
        return s;
    }

    [Fact]
    public void OutputSettingsToEncode_OriginalPresetIsStreamCopy()
    {
        var r = Encode.OutputSettingsToEncode(Settings(s => s.QualityPreset = QualityPreset.Original), Media);
        Assert.False(r.Reencode);
    }

    [Fact]
    public void OutputSettingsToEncode_NamedPresetsMapToCrfAndForceReencode()
    {
        Assert.Equal(18, Encode.OutputSettingsToEncode(Settings(s => s.QualityPreset = QualityPreset.High), Media).Video.Crf);
        Assert.Equal(23, Encode.OutputSettingsToEncode(Settings(s => s.QualityPreset = QualityPreset.Balanced), Media).Video.Crf);
        var small = Encode.OutputSettingsToEncode(Settings(s => s.QualityPreset = QualityPreset.Small), Media);
        Assert.Equal(28, small.Video.Crf);
        Assert.True(small.Reencode);
    }

    [Fact]
    public void OutputSettingsToEncode_BitrateModeSetsBitrateKbps()
    {
        var r = Encode.OutputSettingsToEncode(
            Settings(s => { s.QualityMode = QualityMode.Bitrate; s.VideoBitrateKbps = 5000; }), Media);
        Assert.Equal(5000, r.Video.BitrateKbps);
        Assert.True(r.Reencode);
    }

    [Fact]
    public void OutputSettingsToEncode_VbrMapsToCrfPlusMaxrateCap()
    {
        var r = Encode.OutputSettingsToEncode(
            Settings(s => { s.QualityMode = QualityMode.Vbr; s.Crf = 22; s.VideoBitrateKbps = 9000; }), Media);
        Assert.Equal(22, r.Video.Crf);
        Assert.Equal(9000, r.Video.MaxrateKbps);
        Assert.True(r.Reencode);

        var outArgs = Args(
            new EditSpec { AudioTracks = [Track(0, true, 1)] },
            o => { o.Reencode = true; o.Video = r.Video; });
        var joined = string.Join(" ", outArgs);
        Assert.Contains("-crf 22", joined);
        Assert.Contains("-maxrate 9000k", joined);
        Assert.Contains("-bufsize 18000k", joined);
    }

    [Fact]
    public void OutputSettingsToEncode_FpsMaxHeightClampAgainstSource()
    {
        var up = Encode.OutputSettingsToEncode(Settings(s => { s.Fps = 120; s.MaxHeight = 2160; }), Media);
        Assert.Null(up.Video.Fps);
        Assert.Null(up.Video.ScaleHeight);
        var down = Encode.OutputSettingsToEncode(Settings(s => { s.Fps = 30; s.MaxHeight = 1080; }), Media);
        Assert.Equal(30, down.Video.Fps);
        Assert.Equal(1080, down.Video.ScaleHeight);
    }
}

File: iced/crates/qlipq-ffmpeg/tests/args_tests.rs
use qlipq_core::config::{OutputSettings, QualityMode, QualityPreset};
use qlipq_core::edit_spec::{AudioTrackSpec, CropSpec, EditSpec, TrimSpec};
use qlipq_core::media::MediaInfo;
use qlipq_ffmpeg::args::*;

fn track(index: i64, enabled: bool, volume: f64) -> AudioTrackSpec {
    AudioTrackSpec { index, enabled, volume }
}

fn args(spec: EditSpec, configure: impl FnOnce(&mut BuildExportOptions)) -> Vec<String> {
    let mut opts = BuildExportOptions {
        input_path: "in.mkv".into(),
        output_path: "out.mp4".into(),
        spec,
        reencode: false,
        progress: false,
        video: None,
        audio: None,
        metadata: vec![],
    };
    configure(&mut opts);
    build_export_args(&opts)
}

fn strs(items: &[&str]) -> Vec<String> {
    items.iter().map(|s| s.to_string()).collect()
}

#[test]
fn trim_only_defaults_to_fast_stream_copy() {
    let out = args(
        EditSpec { trim: Some(TrimSpec { start_sec: 5.0, end_sec: 12.5 }), crop: None, audio_tracks: vec![track(0, true, 1.0)] },
        |_| {},
    );
    assert_eq!(
        out,
        strs(&["-y", "-ss", "5.000", "-i", "in.mkv", "-t", "7.500", "-map", "0:v:0", "-map", "0:a:0", "-c:v", "copy", "-c:a", "copy", "out.mp4"])
    );
}

#[test]
fn forced_reencode_reencodes_video_copies_audio() {
    let out = args(
        EditSpec { trim: Some(TrimSpec { start_sec: 0.0, end_sec: 10.0 }), crop: None, audio_tracks: vec![track(0, true, 1.0)] },
        |o| o.reencode = true,
    );
    assert!(out.iter().any(|s| s == "-c:v"));
    assert!(out.iter().any(|s| s == "libx264"));
    assert!(out.join(" ").contains("-c:a copy"));
    assert!(!out.iter().any(|s| s == "-filter_complex"));
}

#[test]
fn crop_builds_filter_and_reencodes() {
    let out = args(
        EditSpec { trim: None, crop: Some(CropSpec { x: 100, y: 50, width: 1280, height: 720 }), audio_tracks: vec![track(0, true, 1.0)] },
        |_| {},
    );
    let j = out.join(" ");
    assert!(j.contains("-filter_complex [0:v:0]crop=1280:720:100:50[vout]"));
    assert!(j.contains("-map [vout]"));
    assert!(j.contains("-map 0:a:0"));
    assert!(j.contains("-c:v libx264"));
    assert!(j.contains("-c:a copy"));
}

#[test]
fn audio_volume_reencodes_audio_copies_video() {
    let out = args(
        EditSpec { trim: None, crop: None, audio_tracks: vec![track(0, true, 0.5), track(1, true, 1.0)] },
        |_| {},
    );
    let j = out.join(" ");
    assert!(j.contains("-filter_complex [0:a:0]volume=0.5[aout0]"));
    assert!(j.contains("-map 0:v:0"));
    assert!(j.contains("-map [aout0]"));
    assert!(j.contains("-map 0:a:1"));
    assert!(j.contains("-c:v copy"));
    assert!(j.contains("-c:a aac -b:a 192k"));
}

#[test]
fn crop_plus_volume_combines_filters() {
    let out = args(
        EditSpec { trim: None, crop: Some(CropSpec { x: 0, y: 0, width: 640, height: 480 }), audio_tracks: vec![track(0, true, 2.0)] },
        |_| {},
    );
    let j = out.join(" ");
    assert!(j.contains("[0:v:0]crop=640:480:0:0[vout];[0:a:0]volume=2[aout0]"));
    assert!(j.contains("-c:v libx264"));
    assert!(j.contains("-c:a aac"));
}

#[test]
fn disabling_all_audio_yields_an() {
    let out = args(
        EditSpec { trim: None, crop: None, audio_tracks: vec![track(0, false, 1.0), track(1, false, 1.0)] },
        |_| {},
    );
    assert!(out.iter().any(|s| s == "-an"));
    assert!(!out.iter().any(|s| s == "-c:a"));
}

#[test]
fn progress_flag_appends_progress() {
    let out = args(EditSpec { trim: None, crop: None, audio_tracks: vec![track(0, true, 1.0)] }, |o| o.progress = true);
    assert!(out.join(" ").contains("-progress pipe:1 -nostats"));
}

#[test]
fn custom_encoder_options() {
    let out = args(
        EditSpec { trim: None, crop: Some(CropSpec { x: 0, y: 0, width: 100, height: 100 }), audio_tracks: vec![track(0, true, 0.0)] },
        |o| {
            o.video = Some(VideoEncodeOptions { codec: Some("libx265".into()), crf: Some(28), preset: Some("fast".into()), ..Default::default() });
            o.audio = Some(AudioEncodeOptions { codec: Some("libopus".into()), bitrate: Some("96k".into()) });
        },
    );
    let j = out.join(" ");
    assert!(j.contains("-c:v libx265 -preset fast -crf 28"));
    assert!(j.contains("-c:a libopus -b:a 96k"));
}

#[test]
fn metadata_stamps_before_output() {
    let out = args(EditSpec { trim: None, crop: None, audio_tracks: vec![track(0, true, 1.0)] }, |o| {
        o.metadata = vec![("game".into(), "Deadlock".into())];
    });
    assert!(out.join(" ").contains("-metadata game=Deadlock"));
    let md = out.iter().position(|s| s == "-metadata").unwrap();
    let outp = out.iter().position(|s| s == "out.mp4").unwrap();
    assert!(md < outp);
}

#[test]
fn frame_rate_change_emits_r_without_filter() {
    let out = args(EditSpec { trim: None, crop: None, audio_tracks: vec![track(0, true, 1.0)] }, |o| {
        o.video = Some(VideoEncodeOptions { fps: Some(30), ..Default::default() });
    });
    let j = out.join(" ");
    assert!(j.contains("-c:v libx264"));
    assert!(j.contains("-r 30"));
    assert!(!out.iter().any(|s| s == "-filter_complex"));
    assert!(j.contains("-c:a copy"));
}

#[test]
fn downscale_builds_scale_filter() {
    let out = args(EditSpec { trim: None, crop: None, audio_tracks: vec![track(0, true, 1.0)] }, |o| {
        o.video = Some(VideoEncodeOptions { scale_height: Some(720), ..Default::default() });
    });
    let j = out.join(" ");
    assert!(j.contains("-filter_complex [0:v:0]scale=-2:720[vout]"));
    assert!(j.contains("-map [vout]"));
    assert!(j.contains("-c:v libx264"));
}

#[test]
fn crop_and_downscale_compose() {
    let out = args(
        EditSpec { trim: None, crop: Some(CropSpec { x: 0, y: 0, width: 1920, height: 1080 }), audio_tracks: vec![] },
        |o| o.video = Some(VideoEncodeOptions { scale_height: Some(720), ..Default::default() }),
    );
    assert!(out.join(" ").contains("[0:v:0]crop=1920:1080:0:0,scale=-2:720[vout]"));
}

#[test]
fn bitrate_uses_bv_not_crf() {
    let out = args(EditSpec { trim: None, crop: None, audio_tracks: vec![track(0, true, 1.0)] }, |o| {
        o.reencode = true;
        o.video = Some(VideoEncodeOptions { bitrate_kbps: Some(6000), ..Default::default() });
    });
    let j = out.join(" ");
    assert!(j.contains("-b:v 6000k"));
    assert!(!j.contains("-crf"));
}

fn media() -> MediaInfo {
    MediaInfo { duration_sec: 60.0, width: 2560, height: 1440, video_codec: "h264".into(), fps: 60.0, audio_streams: vec![], size_bytes: None }
}

fn settings(configure: impl FnOnce(&mut OutputSettings)) -> OutputSettings {
    let mut s = OutputSettings::default();
    configure(&mut s);
    s
}

#[test]
fn encode_original_is_stream_copy() {
    let r = output_settings_to_encode(&settings(|s| s.quality_preset = QualityPreset::Original), &media());
    assert!(!r.reencode);
}

#[test]
fn encode_named_presets_map_to_crf() {
    assert_eq!(output_settings_to_encode(&settings(|s| s.quality_preset = QualityPreset::High), &media()).video.crf, Some(18));
    assert_eq!(output_settings_to_encode(&settings(|s| s.quality_preset = QualityPreset::Balanced), &media()).video.crf, Some(23));
    let small = output_settings_to_encode(&settings(|s| s.quality_preset = QualityPreset::Small), &media());
    assert_eq!(small.video.crf, Some(28));
    assert!(small.reencode);
}

#[test]
fn encode_bitrate_mode() {
    let r = output_settings_to_encode(&settings(|s| { s.quality_mode = QualityMode::Bitrate; s.video_bitrate_kbps = 5000; }), &media());
    assert_eq!(r.video.bitrate_kbps, Some(5000));
    assert!(r.reencode);
}

#[test]
fn encode_vbr_maps_to_crf_plus_maxrate() {
    let r = output_settings_to_encode(&settings(|s| { s.quality_mode = QualityMode::Vbr; s.crf = 22; s.video_bitrate_kbps = 9000; }), &media());
    assert_eq!(r.video.crf, Some(22));
    assert_eq!(r.video.maxrate_kbps, Some(9000));
    assert!(r.reencode);

    let out = args(EditSpec { trim: None, crop: None, audio_tracks: vec![track(0, true, 1.0)] }, |o| {
        o.reencode = true;
        o.video = Some(r.video.clone());
    });
    let j = out.join(" ");
    assert!(j.contains("-crf 22"));
    assert!(j.contains("-maxrate 9000k"));
    assert!(j.contains("-bufsize 18000k"));
}

#[test]
fn encode_fps_maxheight_clamp_against_source() {
    let up = output_settings_to_encode(&settings(|s| { s.fps = 120; s.max_height = 2160; }), &media());
    assert_eq!(up.video.fps, None);
    assert_eq!(up.video.scale_height, None);
    let down = output_settings_to_encode(&settings(|s| { s.fps = 30; s.max_height = 1080; }), &media());
    assert_eq!(down.video.fps, Some(30));
    assert_eq!(down.video.scale_height, Some(1080));
}
